src/main/scala/mesosphere/marathon/state/AppDefinition.scala

```diff
@@ -237,17 +259,80 @@ case class AppDefinition(
     mergeFromProto(proto)
   }
 
-  def withNormalizedVersion: AppDefinition = copy(version = Timestamp(0))
+  def version: Timestamp = versionInfo.version
 
+  /**
+    * Returns whether this is a scaling change only.
+    */
   def isOnlyScaleChange(to: AppDefinition): Boolean =
     !isUpgrade(to) && (instances != to.instances)
 
   def isUpgrade(to: AppDefinition): Boolean =
-    this != to.copy(instances = instances, version = version)
+    this != to.copy(instances = instances, versionInfo = versionInfo)
+
+  /**
+    * Returns the changed app definition that is marked for restarting.
+    */
+  def markedForRestarting: AppDefinition = copy(versionInfo = VersionInfo.NoVersion)
+
+  /**
+    * Returns true if we need to restart all tasks.
+    *
+    * This can either be caused by changed configuration (e.g. a new cmd, a new docker image version)
+    * or by a forced restart.
+    */
+  def needsRestart(to: AppDefinition): Boolean =
+    to.versionInfo == AppDefinition.VersionInfo.NoVersion || isUpgrade(to)
+
 }
 
 object AppDefinition {
 
+  sealed trait VersionInfo {
+    def version: Timestamp
+
+    def withScaleOrRestartChange(newVersion: Timestamp): VersionInfo = {
+      VersionInfo.forNewConfig(version).withScaleOrRestartChange(newVersion)
+    }
+
+    def withConfigChange(newVersion: Timestamp): VersionInfo = {
+      VersionInfo.forNewConfig(newVersion)
+    }
+  }
+
+  object VersionInfo {
+    case object NoVersion extends VersionInfo {
+      def version: Timestamp = Timestamp(0)
+    }
+
+    /**
+      * Only contains a version timestamp. Will be converted to a FullVersionInfo before stored.
+      */
+    case class OnlyVersion(version: Timestamp) extends VersionInfo
+
+    /**
+      * @param version The versioning timestamp (we are currently assuming that this is the same as lastChangeAt)
+      * @param lastScalingAt The time stamp of the last change including scaling or restart changes
+      * @param lastConfigChangeAt The time stamp of the last change that changed configuration
+      *                               besides scaling or restarting
+      */
+    case class FullVersionInfo(
```

Comment:
minor nit-picking, but since it's about naming I'd like to share my thoughts:
What I like about `NoVersion` and `OnlyVersion` is, that both make clear, that these objects are somewhat exceptional/corner cases. However, `FullVersionInfo`, besides being the only one with `Info` suffix, _sounds_ like it's somewhat _unusual_ because of the `full` prefix, which is not the case. I currently have no better proposal at hand, but if you agree, let's have a quick chat about a nicer name :)